### ISSUE TITLE
[ErrorHandler] : Throw exception on notice in debug mode

### ISSUE THREAD

**Issue by QuentinCurtet** [symfony/symfony#35336]
**Symfony version(s) affected**: 5.0.2 (and prior i guess)
**Description**  
A notice (for example : "Notice: Undefined variable: index"), throws an exception in debug mode in dev env.

**How to reproduce**  
Create a basic Command file TestCommand.php
Set env variables like this :
* dev => true
* debug => true
```
protected function configure() {
	$this->setName('app:test');
}
/**
 * @param InputInterface $input
 * @param OutputInterface $output
 * @return int
 */
protected function execute(InputInterface $input, OutputInterface $output) {
	echo $index;
	echo 'toto';
	return 0;
}
```
Launch the following command from terminal :
`php bin/console app:test`

It will throw an exception in console and not show 'toto'.

**Possible Solution**  
In debug mode, if it's not an error or higher (critical, fatal adn so on), it doesn't have to throw an exception, but only display a message and continue the execution of the command.

**Additional context**  
![image](https://user-images.githubusercontent.com/6769513/72344325-3c8a9000-36d1-11ea-801b-5a0114989742.png)

**Comment by stephen-lewis** [comment]
I'm not sure a notice should throw and exception - the definition of E_NOTICE in the php docs is:

> Run-time notices. Indicate that the script encountered something that could indicate an error, but could also happen in the normal course of running a script.

If a notice threw an exception it would potentially break otherwise working/correct code. I've seen notices generated from bad config in php.ini when running good code.

If this is required, perhaps a flag to enable exceptions on notice would be better.

**Comment by nicolas-grekas** [comment]
Not accepting a single deprecation notice during dev is a best practice that is enforced since the first days of Symfony. It encourages writing better code. That's desired.

**Comment by QuentinCurtet** [comment]
There is a difference between display a weel formed message and throw an exception that kill the process...
Is there a way to disable it while i'm in debug mode ?

**Comment by QuentinCurtet** [comment]
@nicolas-grekas I'm ok with the fact to write better code, but i think symfony should let the choice to throw exception or not for a notice.
At least can you tell me in which file this part is coded so i can take a look ?

Thanks.

**Comment by derrabus** [comment]
I feel like we‘ve had that discussion in #34968 already.

**Comment by nicolas-grekas** [comment]
Note that you can use the `@` silencing operator when you do expect a php notice or warning but decide to ignore it. That's the case eg. when calling some php functions like `fopen()`.
But silently ignoring the use of undefined variables or undefined indexes is a bad code smell that these exceptions prevent from happening.

In which situation do you want to keep getting notices? Does the silencing operator solve them?

**Comment by QuentinCurtet** [comment]
@derrabus Last time, something on the doc (migration to 4.4) was not clearly defined.
This time it's a little bit different.

@nicolas-grekas It's not a problem to ignore notice, the real problem is that a php notice should not throw any exception (no matter what the env or debug is).

Is there a possibility to fix this without using `@` ?

**Comment by nicolas-grekas** [comment]
> the real problem is that a php notice should not throw any exception

I'm sorry but this is not how we're going to make progress in this discussion. The majority here will think that an exception should be thrown when a PHP notice happen, I can bet on it.

**Comment by QuentinCurtet** [comment]
Throw a poll ;)
Futhermore, why create a difference between dev and other env (in production mode, the notice will not throw any exception am i wrong ?)

**Comment by nicolas-grekas** [comment]
What problem are you trying to solve? From a design pov, the current settings are proven and are not going to change, unless we can discuss actual use cases where we'd have some concrete food for thoughts. But if we're going to have an abstract discussion, things are as desired.

**Comment by xabbuh** [comment]
By the way, this has been the Symfony philosophy since years. Being more strict during development to enable the developer to code as carefully as possible while not throwing on everything in production to not influence the user's experience too badly (see the behaviour for unknown variables in Twig templates for example).

**Comment by QuentinCurtet** [comment]
For example, on data you don't have hand on (like a file provided by a supplier).
You don't always know what the data format is.
When you try to integrate those data into your system, it can display notices because some data are missing.
In this case, you must show the notice, but never throw an exception that can kill the process.

Aren't you agree ?

**Comment by QuentinCurtet** [comment]
In order to discuss in more constructive way :
First of all, we have the same philosophy of code than you, we always try to fix notices/warnings and all levels of php errors in order to have a better code quality.
Of course, we understand that, and we approve ! This is what we are doing every day with our developpers colleagues.
But, in some situation we unfortunately experienced, this rule for doing better code can become a constraint, and I will try to explain this with the example below.

We have some code based on external data (ie : some provider give us a file with data to integrate).
But in production environment, sometimes, the external supplier put wrong data in the file that raise php notices (unpredictable data or format changed without reason...).
When we try to fix the problem in dev environment (with debug mode we would like to keep), we are totally blocked by the fact that notices are treated as exception.
So in this case, it would be nice if we'll be able to reproduce the issue without throwing exception (and without disabling the debug mode).
This will really help us to be more efficient for better understanding what's wrong in the data, and to adapt the code.
It may be, for example, a config option to choose the error level to throw exception at.

We hope this will help to better understand what we mean, and we also hope that you'll have to consider these situations for having facilities to reproduce, debug and solve in development environment the notice error we see in production.
Please, let us know what you think.

**Comment by nicolas-grekas** [comment]
@QuentinCurtet simply: nope :) No parsing logic should ever throw notices. Parsers should be written in a way that doesn't throw any of them - that's totally possible, and that leads to better code.

I'm going to close this PR because this is now a support issue, nothing to keep track of as far as the code in this repo is concerned. Thanks for asking.

**Comment by mask8** [comment]
I understand it's good practice to eliminate any notice level errors, but when this notice occurs in third-party library, doctrine for instance, having no control to stop this exception is very annoying.

**Comment by jameslmoser** [comment]
> I understand it's good practice to eliminate any notice level errors, but when this notice occurs in third-party library, doctrine for instance, having no control to stop this exception is very annoying.

I was able to resolve this by changing this to false:
https://symfony.com/doc/current/reference/configuration/framework.html#throw

Hope that helps.

**Comment by QuentinCurtet** [comment]
> > I understand it's good practice to eliminate any notice level errors, but when this notice occurs in third-party library, doctrine for instance, having no control to stop this exception is very annoying.
> 
> I was able to resolve this by changing this to false:
> https://symfony.com/doc/current/reference/configuration/framework.html#throw
> 
> Hope that helps.

You're link helped me to find my solution.
I only add the parameter debug.error_handler.throw_at to 0 in order to have the same behavior between all my envs.

Thanks @jameslmoser

**Comment by php4fan** [comment]
I get that this is the correct _default_ behavior, but what is the recommended clean way to change it?? If there isn't one, then that's wrong.

Say I have some legacy code that is full of uncaught notices, that I have "migrated" to Symfony. I have to postpone fixing that to the future; I just can't afford focusing on such level of detail right now. Meanwhile, I need to just be able to run my application. So how do I tell Symfony to not stop execution and throw an exception below a certain level (e.g. <= E_NOTICE) even in dev environment (I'm assuming this doesn't happen in prod)?
I still want the notices to be logged BTW.

Outside of Symfony, I would just change `error_reporting`; except that would also prevent me from logging such issues.
